### PR TITLE
Restructured code and Added PINNs-iVFM

### PR DESCRIPTION
**Pull request by HangJung97**
## What does this PR do?

Restructured code and Added PINNs-iVFM.

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **run pre-commit hooks** with `pre-commit run -a` command?

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/tasks/ivfm.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    torch.cuda.manual_seed(12345)
+
+# Create data
+# Set current working directory
+cwd = Path(__file__).resolve().parent
+os.chdir(cwd)
+patient = "../../data/patient_06_1.mat"
+data = loadmat(patient)
+fname = os.path.basename(patient)[:-4]
+spacing = [0.5, 0.5, 1]
```
Comment (olivier-bernard-creatis) [inline src/tasks/ivfm.py:30]: Are we sure that the spacing is the same for each data / patient ? Wouldn't it be better to put this information in the ".mat" file and load it here ?

File: src/tasks/ivfm.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+# Weights for losses
+lambda_div = 1e-4
+lambda_BC = 1
+lambda_smooth = 1e-5
+
+nb_frame = data["uR"].shape[-1]
+preds_array = np.zeros([2, *data["uR"].shape])
+
+# nb_frame = 1
```
Comment (olivier-bernard-creatis) [inline src/tasks/ivfm.py]: To remove

File: src/tasks/ivfm.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+# Create an MLP network
+mlp = MLP([2], [2], [100, 100, 100, 100, 100, 100, 100, 100], activation="tanh")
+
+# Define optimizer, learning rate, and number of epochs
+lr = 1e-4
+optimizer = partial(torch.optim.Adam, lr=lr)
+# optimizer = partial(torch.optim.LBFGS, lr=2, max_iter=100000)
```
Comment (olivier-bernard-creatis) [inline src/tasks/ivfm.py:53]: should we remove this comment line ?

File: src/tasks/ivfm.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    lambda_div=lambda_div,
+    lambda_BC=lambda_BC,
+    lambda_smooth=lambda_smooth,
+).to(DEVICE)
+
+nb_frame = 1
+for frame in range(nb_frame):
+    # Get radial and angular velocity
+    VNyquist = data["VNyquist"]
+    powers = np.log10(data["powers"][..., frame]) / 2
```
Comment (olivier-bernard-creatis) [inline src/tasks/ivfm.py:71]: Why do you divide the power by 2 ?

File: src/models/pinns_ivfm.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        """Compute physical loss."""
+        grid_R = self.np_to_th(batch["grid_R"], device).requires_grad_(True)
+        grid_TH = self.np_to_th(batch["grid_TH"], device).requires_grad_(True)
+        # uD = self.np_to_th(batch["uD"], device)
+
+        x = torch.hstack((grid_R, grid_TH))
+        # x = torch.hstack((grid_R, grid_TH, uD))
+
+        # Compute the corresponding estimated velocities
+        preds = self.forward(x)
```
Comment (olivier-bernard-creatis) [inline src/models/pinns_ivfm.py:81]: I just wondering the correctness of applying two forward passes consecutively, one for the data attachment loss and one here for the physical loss. This is not a comment to change the code, but more a generic question to be discussed in our upcoming meetings ;)

File: src/models/pinns_ivfm.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        uR = preds[:, :-1]
```
Comment (olivier-bernard-creatis) [inline src/models/pinns_ivfm.py:82]: is there an interest not to put here "uR = preds[:, 0]" and "uTH = preds[:, 1]" ?

**Review by olivier-bernard-creatis — CHANGES_REQUESTED** [review]
Excellent code Hang Jung, congrats ;)

File: src/tasks/ivfm.py (hunk 1 above)
Comment (HangJung97) [reply src/tasks/ivfm.py:30]: Spacing is only used for saving the nifti file and for illustration purpose when the file is opened in ITK snap. Since we did not interpolate the image, the spacing is actually “useless” and won't affect the results.

File: src/tasks/ivfm.py (hunk 2 above)
Comment (HangJung97) [reply src/tasks/ivfm.py]: Indeed!

File: src/tasks/ivfm.py (hunk 3 above)
Comment (HangJung97) [reply src/tasks/ivfm.py:53]: This is an optimizer that I want to try, since a lot of PINNs used the L-BFGS optimizer. However, it would require some code refactoring. To be discussed...

File: src/tasks/ivfm.py (hunk 4 above)
Comment (HangJung97) [reply src/tasks/ivfm.py:71]: To normalize the power between 0 and 1.

File: src/models/pinns_ivfm.py (hunk 5 above)
Comment (HangJung97) [reply src/models/pinns_ivfm.py:81]: A brief explanation: normally in PINNs, the input (measurements of which we have the reference) used for computing the reconstruction loss during the usual forward pass is different from the one (some random points in the domain) used for computing the physical loss.

File: src/models/pinns_ivfm.py (hunk 6 above)
Comment (HangJung97) [reply src/models/pinns_ivfm.py:82]: To be consistent with the input of size (N, 1).

File: src/tasks/ivfm.py (hunk 1 above)
Comment (olivier-bernard-creatis) [reply src/tasks/ivfm.py:30]: This is true if all the native image are isotropic, is it the case ?

File: src/tasks/ivfm.py (hunk 3 above)
Comment (olivier-bernard-creatis) [reply src/tasks/ivfm.py:53]: OK

File: src/tasks/ivfm.py (hunk 4 above)
Comment (olivier-bernard-creatis) [reply src/tasks/ivfm.py:71]: OK, maybe it would be nice to add comments more comments on the code to help to understand this type of things ;)

File: src/models/pinns_ivfm.py (hunk 5 above)
Comment (olivier-bernard-creatis) [reply src/models/pinns_ivfm.py:81]: This I understand, but why not using only one single forward pass ? any idea ? To be discussed ;)

File: src/models/pinns_ivfm.py (hunk 6 above)
Comment (olivier-bernard-creatis) [reply src/models/pinns_ivfm.py:82]: Not sure to understand, to be discussed ;)

**Review by olivier-bernard-creatis — APPROVED** [review]
I let you do a last commit with a few lines of code to clean and with more comments on key parts of the code ;))))
